cmd/disk_wipe.go

```diff
@@ -0,0 +1,93 @@
+package cmd
+
+import (
+	"cmp"
+	"context"
+	"errors"
+	"os"
+
+	"github.com/bmc-toolbox/common"
+	"github.com/metal-toolbox/ironlib"
+	"github.com/metal-toolbox/ironlib/actions"
+	"github.com/metal-toolbox/ironlib/utils"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+func init() {
+	cmd := &cobra.Command{
+		Use:   "wipe /dev/disk",
+		Short: "Wipes all data from a disk",
+		Args: func(_ *cobra.Command, args []string) error {
+			if len(args) < 1 {
+				return errors.New("requires at least one arg") // nolint:goerr113
+			}
+
+			_, err := os.Open(args[0])
+			return err
+		},
+		Run: func(cmd *cobra.Command, args []string) {
+			timeout, err := cmd.Flags().GetDuration("timeout")
+			if err != nil {
+				logger.With("error", err).Fatal("--timeout argument is invalid")
+			}
+
+			verbose, err := cmd.Flags().GetBool("verbose")
+			if err != nil {
+				logger.With("error", err).Fatal("--verbose argument is invalid")
+			}
+
+			driveName := args[0]
+
+			logger := logrus.New()
+			logger.Formatter = new(logrus.TextFormatter)
+			if verbose {
+				logger.SetLevel(logrus.TraceLevel)
+			}
+			l := logger.WithField("drive", driveName)
+
+			ctx := cmp.Or(cmd.Context(), context.Background())
+			ctx, cancel := context.WithTimeout(ctx, timeout)
+			defer cancel()
+
+			collector, err := ironlib.New(logger)
+			if err != nil {
+				l.WithError(err).Fatal("exiting")
+			}
+
+			inventory, err := collector.GetInventory(ctx, actions.WithDynamicCollection())
+			if err != nil {
+				l.WithError(err).Fatal("exiting")
+			}
+
+			var drive *common.Drive
+			for _, d := range inventory.Drives {
+				if d.LogicalName == driveName {
+					drive = d
+					break
+				}
+			}
+			if drive == nil {
+				l.Fatal("unable to find disk")
+			}
+
+			// Pick the most appropriate wipe based on the disk type and/or features supported
+			var wiper actions.DriveWiper
+			switch drive.Protocol {
+			case "nvme":
```

Comment:
I imagine the intention here is to add more cases in the future?

Reply:
Also, shouldn't this 'figure out what tool to use based upon protocol' be moved into ironlib?

Reply:
Yep but I’d want the Drives to support that themselves otherwise instead of having a function that needs to grow more and more complex as features are needed. In that case ironlib discovers the drives that know how to wipe themselves and vogelkop chooses which ones depending on cli args maybe. But that’s a v2 thing.

Reply:
I feel like scattering this logic everywhere that needs it is wrong...    Maybe there should be another lib in the middle that contains the logic you'll otherwise implement in vogelkop.    I imagine reusing this logic elsewhere...  but maybe ultimately nothing else will ever use it...

Reply:
I _think_ what you're both talking about here is the distinction between implementation and behaviour.  Implementation, for instance, might vary from one disk to the next: what we do to a disk is dependent on the disk type, and given the disk type is an interface we would want a call that reflects the _behaviour_ which is diskX.destroy_all_data_with_prejudice() or diskX.lightly_mangle_data() or diskX.do_what_you_can_in_5_minutes().

Reply:
The point about hiding the implementation for a disk type is that I should be able to test that the implementation for a disk type produces a result without having to know _how_ it does that.

Reply:
> I _think_ what you're both talking about here is the distinction between implementation and behaviour. Implementation, for instance, might vary from one disk to the next: what we do to a disk is dependent on the disk type, and given the disk type is an interface we would want a call that reflects the _behaviour_ which is diskX.destroy_all_data_with_prejudice() or diskX.lightly_mangle_data() or diskX.do_what_you_can_in_5_minutes().

Not quite because the disk type is not an interface and has none, its just a POD struct. So we either provided a function to take the drive and return back some behavior or we make the drive implement an interface so that we can leave that to the disk or we let caller deal with it. I vote to let the caller deal with it instead of providing the function and later implement interfacification in the drive :D